Change PowerDistributor to ignore components with NaN metrics

ela-kotulska-frequenz opened the pull request:
Each float metric returned from api can be float("NaN"). If that happen PowerDistributor would fail because it is impossible to make math calcuation on NaN value.
This PR is just a quick fix of this problem.
* if power bounds are NaN, then it tries to replace them with corresponding power bounds from adjacent component. If these components are also NaN, then it ignores battery.
* if other metrics metrics are NaN then it ignores battery.

More sophisticated solution should be done in the future, maybe use resampler.

ela-kotulska-frequenz committed:
Change PowerDistributor to ignore components with NaN metrics

Each float metric returned from api can be float("NaN").
If that happen PowerDistributor would fail because it is
impossible to make math calcuation on NaN value.
This PR is just a quick fix of this problem.

* if power bounds are NaN, then it tries to replace them
with corresponding power bounds from adjacent component.
If these components are also NaN, then it ignores battery.

* if other metrics metrics are NaN then it ignores battery.

More sophisticated solution should be done in the future.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -558,25 +558,76 @@ def _get_components_data(self, batteries: Set[int]) -> List[InvBatPair]:
+
+        # It means that nothing has been send on this channels, yet.
+        # This should be handled by BatteryStatus so it should never happen.
+        if battery_data is None or inverter_data is None:
```
Not sure if it's worth it, but if it should never happen maybe log an error in case there is some problem that can be detected earlier.

ela-kotulska-frequenz replied:
This moment when you get `Not sure if it's worth it` comment and you realize it is bug :D
https://github.com/frequenz-floss/frequenz-sdk-python/pull/247/commits/198f05f71a3607cfa496133813c7f79caebaf600

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -558,25 +558,76 @@ def _get_components_data(self, batteries: Set[int]) -> List[InvBatPair]:
+            battery_data.capacity,
+        ]
+        if any(map(isnan, not_replaceable_metrics)):
+            _logger.warning("Some metrics for battery %d are NaN", battery_id)
```
Not sure if this should be a warning, couldn't it spam the logs too much if some device is sending `NaN`s for a while? Isn't this sort of *normal* behavior?

ela-kotulska-frequenz replied:
Done, You are right. I changed it to debug log

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -558,25 +558,76 @@ def _get_components_data(self, batteries: Set[int]) -> List[InvBatPair]:
+        ]
+
+        # If all values are ok then return them
+        if all(map(lambda x: not isnan(x), replaceable_metrics)):
```
Just for fun :D, this could also be written:
```py
         if not any(map(isnan, replaceable_metrics)):
```

ela-kotulska-frequenz replied:
Haha Thanks, done :D

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -558,25 +558,76 @@ def _get_components_data(self, batteries: Set[int]) -> List[InvBatPair]:
+            pairs_data.append(data)
         return pairs_data
 
+    def _get_battery_inverter_data(
```
Maybe it would be nice to have a comment on this method even if it's private to mention that this is dealing mainly with `NaN`s and how they are being handled?

ela-kotulska-frequenz replied:
Described

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -558,25 +558,76 @@ def _get_components_data(self, batteries: Set[int]) -> List[InvBatPair]:
+            bat_bound = getattr(battery_data, bat_attr)
+            inv_bound = getattr(inverter_data, inv_attr)
+            if isnan(bat_bound) and isnan(inv_bound):
+                _logger.warning("Some metrics for battery %d are NaN", battery_id)
```
Same comment about confirming if we really want this to be a warning.

ela-kotulska-frequenz replied:
We don't. Changed to debug.

leandro-lucarella-frequenz reviewed (dismissed):
I didn't look into the tests in depth, and other than the small comments, LGTM. They are all minor, so I leave it to you if you want to change any of them, so giving an approval.

ela-kotulska-frequenz committed:
Change `get_components_data` to get data for working batteries only

This method is also used for preprocessing the request to check if
request is correct before adding it to the main task queue.
In methods: `_get_upper_bound` and `get_lower_bound`.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -558,24 +558,100 @@ def _get_components_data(self, batteries: Set[int]) -> List[InvBatPair]:
-            pairs_data.append(InvBatPair(battery_data, inverter_data))
-        return pairs_data
+        Each float data from the microgrid can be "NaN".
+        We can't do math operations on "NaN".
+        So check all the metrics and:
+        * if power bounds are NaN, then try to replace it with the corresponding
+        power bounds from the adjacent component. If metric in the adjacent component is
+        also NaN, then return None.
+        * if other metrics are NaN then return None. We can't assume anything for other
+        metrics.
```
Not important as this is private and it won't get rendered, but just a reminder for the cases where it will:

```suggestion
        We can't do math operations on "NaN".        
        So check all the metrics and:
        * if power bounds are NaN, then try to replace it with the corresponding
          power bounds from the adjacent component. If metric in the adjacent component is
          also NaN, then return None.
        * if other metrics are NaN then return None. We can't assume anything for other
          metrics.
```

ela-kotulska-frequenz replied:
Done

ela-kotulska-frequenz commented:
Rebased and changed formatting.

ela-kotulska-frequenz commented:
> I didn't look into the tests in depth

Tests are copy-paste. :( 
I am not proud of them but it was changed so many times... I hope they would be improved when we do PowerManger.
